beacon_node/eth1_chain/src/cache.rs

```diff
@@ -0,0 +1,182 @@
+use crate::types::Eth1DataFetcher;
+use parking_lot::RwLock;
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use tokio;
+use types::*;
+use web3::futures::*;
+use web3::types::*;
+
+/// Cache for recent Eth1Data fetched from the Eth1 chain.
+#[derive(Clone, Debug)]
+pub struct Eth1DataCache<F: Eth1DataFetcher> {
+    cache: Arc<RwLock<BTreeMap<U256, Eth1Data>>>,
+    last_block: Arc<RwLock<u64>>,
+    fetcher: Arc<F>,
+}
+
+impl<F: Eth1DataFetcher> Eth1DataCache<F> {
+    pub fn new(fetcher: Arc<F>) -> Self {
+        Eth1DataCache {
+            cache: Arc::new(RwLock::new(BTreeMap::new())),
+            // Note: Should ideally start from block where Eth1 chain starts accepting deposits.
+            last_block: Arc::new(RwLock::new(0)),
+            fetcher: fetcher,
+        }
+    }
+
+    /// Called periodically to populate the cache with Eth1Data from most recent blocks.
+    pub fn update_cache(&self) -> impl Future<Item = (), Error = ()> + Send {
+        let cache_updated = self.cache.clone();
+        let last_block = self.last_block.clone();
+        let fetcher = self.fetcher.clone();
+        let my_future = self
+            .fetcher
+            .get_current_block_number()
+            .and_then(move |curr_block_number| {
+                fetch_eth1_data_in_range(0, 10, curr_block_number, fetcher)
+                    .for_each(move |data| {
+                        let data = data.unwrap();
+                        let mut eth1_cache = cache_updated.write();
+                        eth1_cache.insert(data.0, data.1);
+                        Ok(())
+                    })
+                    .map_err(|e| println!("Fetching in range failed {:?}", e))
+                    .and_then(move |_| {
+                        let mut last_block_updated = last_block.write();
+                        *last_block_updated = curr_block_number.as_u64();
+                        // TODO: Delete older stuff
+                        Ok(())
+                    })
+            })
+            .map_err(|e| println!("Reading current block number failed failed {:?}", e));
+        my_future
+    }
+
+    /// Get `Eth1Data` object at a distance of `distance` from the perceived head of the currrent Eth1 chain.
+    /// Returns the object from the cache if present, else fetches from Eth1Fetcher.
+    pub fn get_eth1_data(&mut self, distance: u64) -> Option<Eth1Data> {
+        let current_block_number: U256 = self.fetcher.get_current_block_number().wait().ok()?;
+        let block_number: U256 = current_block_number.checked_sub(distance.into())?;
+        if self.cache.read().contains_key(&block_number) {
```

Comment:
```suggestion
        if let Some(result) = self.cache.read().get(&block_number) {
        return Some(result);
        else {
```
Something like this simplifies this

Reply:
This is definitely better. Will fix it.